### PR TITLE
EIP: PoS blockhash backward compatibility

### PR DESCRIPTION
**Pull request by beether**
PoW `blockhash` and PoS `blockhash` differ, in that PoW's `blockhash` is economically staked and PoS's `blockhash` is not.

### CODE REVIEW

**Comment by jamesray1** [comment]
This looks OK to me, although it would be good to see some tests.

**Comment by beether** [comment]
@jamesray1 Thanks for digging around, it's much appreciated.

From the paper:

> 12.2. Random Numbers. Providing random numbers
> within a deterministic system is, naturally, an impossible
> task. However, we can approximate with pseudo-random
> numbers by utilising data which is generally unknowable
> at the time of transacting. Such data might include the
> block’s hash, the block’s timestamp and the block’s benefi-
> ciary address. In order to make it hard for malicious miner
> to control those values, one should use the BLOCKHASH
> operation in order to use hashes of the previous 256 blocks
> as pseudo-random numbers.

It looks to me like there is definitely some intent on having Blockhash be `generally unknowable at the time of transacting`. Hopefully we can preserve this in PoS, or provide an alternative.

**Comment by beether** [comment]
@jamesray1 Any progress on this? If there's anything I can do, please let me know.

**Comment by jamesray1** [comment]
Sorry I forgot about this. Just flagging that I'm not an EIP editor. I remember the discussion you had on Gitter with @arachnid, it seemed like the blockhash isn't a reliable source of randomness, you can use something else like BLS random beacons which may be included in the protocol later with Casper in phase 5 of sharding.

**Comment by beether** [comment]
@jamesray1 

> it seemed like the blockhash isn't a reliable source of randomness

It _is_ a reliable source of randomness, with limitations. This EIP seeks to maintain this compatibility.

The importance here is that blockhash _already_ has the characteristics mentioned in the EIP -- namely, it requires an economic incentive to be made _unreliable_.  For backwards compatibility, it should be kept this way.

Aside from backwards compatibility, there is immense value in having an economically staked source of randomness (eg, current blockhash) that is natively accessible to contracts on the blockchain. Enforcing that such a feature is used "correctly" would be another discussion.

**Comment by jamesray1** [comment]
I think this needs further research. Alternative randomness models to the blockhash could also be used.

Why do you want to keep an economic incentive to make the blockhash unreliable? Is that what you meant to say?

**Comment by jamesray1** [comment]
Alternative models include RANDAO (which I think @vbuterin prefers), pure private randomness. Some cryptographers assert that random oracle models are not secure.

**Comment by jamesray1** [comment]
https://vitalik.ca/files/randomness.html
https://blog.cryptographyengineering.com/2011/09/29/what-is-random-oracle-model-and-why-3/
Also Casper won't be implemented until about 5 years from now with phase 5 of Ethereum 2.0, according to the latest roadmap. https://ethresear.ch/t/sharding-phase-1-spec/1407/19

I think you should cross post this on https://ethresear.ch.

**Comment by jamesray1** [comment]
To be honest I don't want to think about this too much as I have enough to do with developing a sharding implementinon.

**Comment by jamesray1** [comment]
Note that the randomness model for block proposers with full PoS CBC Casper has not been decided upon yet, but it is true that using the blockhash as a source of randomness is an option.

**Comment by jamesray1** [comment]
Without a proposed spec and implementation for the randomness model included in this EIP, I think this EIP should be treated as a topic for further research, development and discussion.

For BLS random beacons, they are apparently not as favourable (according to Vitalik), since they are prone to 51% attacks, whereas RANDAO and pure private randomness are not.

**Comment by beether** [comment]
> Why do you want to keep an economic incentive to make the blockhash unreliable? Is that what you meant to say?

I want a source of randomness on the blockchain, that has a predictable cost to meddle with. Currently, blockhash provides this: A miner will not try to create a blockhash with certain characteristics unless given some predictable amount of financial incentive to do so, otherwise they are wasting money by mining uncles instead of winning blocks.

As for the "why?" One answer is for backward compatibility, since blockhash currently provides this. 

Another answer is because it is immensely valuable to have a source of randomness that does not require additional transactions from additional sources -- this assures autonomy for any process that require randomness. BLS beacons, RANDAO, Oracles, and all other schemes require an additional transaction to produce a random number. There is no guarantee a number will ever be produced.

A simple use case of blockhash randomness:  Using a transaction's resultant blockhash to determine if they've won a coin toss, paying 2:1 on a win. _As long as the wager is limited_, there is no incentive for miner's to try to produce an "even" blockhash over an "odd" one.

Another use case may be using blockhash to assure an item is assigned to a random bucket.

Basically, any use case for random numbers where autonomy is important, and where the outcome provides a limited financial incentive for miners to meddle.

**Comment by beether** [comment]
I appreciate you taking the time to discuss this, and sorry to be a thorn in your side. To me, it seems like it could very easily be accommodated in the design of the PoS system and/or sharding -- I personally simply am not familiar enough to write the spec for it.

I'll try to cross-post to ethresearch.

**Comment by Arachnid** [comment]
This is a courtesy notice to let you know that the format for EIPs has been modified slightly. If you want your draft merged, you will need to make some small changes to how your EIP is formatted:

 - Frontmatter is now contained between lines with only a triple dash ('---')
 - Headers in the frontmatter are now lowercase.

If your PR is editing an existing EIP rather than creating a new one, this has already been done for you, and you need only rebase your PR.

In addition, a continuous build has been setup, which will check your PR against the rules for EIP formatting automatically once you update your PR. This build ensures all required headers are present, as well as performing a number of other checks.

Please rebase your PR against the latest master, and edit your PR to use the above format for frontmatter. For convenience, here's a sample header you can copy and adapt:

```
---
eip: <num>
title: <title>
author: <author>
type: [Standards Track|Informational|Meta]
category: [Core|Networking|Interface|ERC] (for type: Standards Track only)
status: Draft
created: <date>
---
```
